File: hanimo_rag/core/judge.py
"""LLM-based relevance judge for search results."""
from __future__ import annotations

import logging

from hanimo_rag.llm.base import LLMBase
from hanimo_rag.types import IndexedChunk

logger = logging.getLogger(__name__)

SYSTEM_PROMPT = (
    "You are a relevance judge. Rate whether each chunk is relevant to a search query. "
    "Always output valid JSON."
)

JUDGE_PROMPT = """Query: {query}

Rate each chunk's relevance to the query. Be strict — only mark as relevant if the chunk
actually helps answer or address the query.

{chunks_formatted}

Output JSON array only:
[{{"id": "chunk_id", "relevant": true}}, {{"id": "chunk_id", "relevant": false}}, ...]"""


class Judge:
    """Filter search results by LLM-judged relevance."""

    def __init__(self, llm: LLMBase) -> None:
        self.llm = llm

    async def filter(
        self,
        query: str,
        chunks: list[IndexedChunk],
        batch_size: int = 10,
    ) -> list[IndexedChunk]:
        """Filter chunks to only those relevant to the query."""
        if not chunks:
            return []

        relevant: list[IndexedChunk] = []
        chunk_map = {c.id: c for c in chunks}

        # Process in batches
        for i in range(0, len(chunks), batch_size):
            batch = chunks[i : i + batch_size]
            batch_relevant = await self._judge_batch(query, batch, chunk_map)
            relevant.extend(batch_relevant)

        return relevant

    async def _judge_batch(
        self,
        query: str,
        batch: list[IndexedChunk],
        chunk_map: dict[str, IndexedChunk],
    ) -> list[IndexedChunk]:
        """Judge a batch of chunks."""
        chunks_formatted = "\n\n".join(
            f"Chunk ID: {c.id}\n"
            f"Summary: {c.summary}\n"
            f"Content preview: {c.content[:300]}"
            for c in batch
        )

        prompt = JUDGE_PROMPT.format(query=query, chunks_formatted=chunks_formatted)
        result = await self.llm.generate_json(prompt, system=SYSTEM_PROMPT)

        relevant: list[IndexedChunk] = []

        # Handle both direct list and dict with "items" key
        judgments: list[dict] = []
        if isinstance(result, dict):
            judgments = result.get("items", [])
            if not judgments:
                # Maybe the result itself has the expected keys
                # This happens when generate_json wraps a list
                for key in result:
                    if isinstance(result[key], list):
                        judgments = result[key]
                        break
        elif isinstance(result, list):
            judgments = result

        if not judgments:
            # Fallback: if LLM failed, return all chunks (be permissive)
            logger.warning("Judge failed to parse results; returning all chunks as relevant")
            return list(batch)

        for j in judgments:
            if not isinstance(j, dict):
                continue
            chunk_id = j.get("id", "")
            is_relevant = j.get("relevant", False)
            if is_relevant and chunk_id in chunk_map:
                relevant.append(chunk_map[chunk_id])

        return relevant

File: js/src/core/judge.ts
/**
 * hanimo-rag v2 — Relevance Judge
 *
 * Uses the LLM to batch-judge whether retrieved chunks are actually
 * relevant to the user's query and assign relevance scores.
 */

import type { LLMBase } from '../llm/base.js'
import type { IndexedChunk, JudgeResult } from '../types.js'

const JUDGE_SYSTEM = `You are a relevance judge. Given a user query and a list of text chunks, rate each chunk's relevance to the query.

Return a JSON array of objects, one per chunk, each with:
- "id": the chunk id (string)
- "relevant": true if the chunk helps answer the query, false otherwise
- "score": relevance score from 0.0 to 1.0 (1.0 = perfectly relevant)

Be strict: only mark chunks as relevant if they genuinely contain information that helps answer the query.`

export class RelevanceJudge {
  private llm: LLMBase

  constructor(llm: LLMBase) {
    this.llm = llm
  }

  /**
   * Judge relevance of chunks against a query.
   * Processes in batches to avoid hitting token limits.
   */
  async judge(
    query: string,
    chunks: IndexedChunk[],
    batchSize: number = 5,
  ): Promise<JudgeResult[]> {
    if (chunks.length === 0) {
      return []
    }

    const results: JudgeResult[] = []

    // Process in batches
    for (let i = 0; i < chunks.length; i += batchSize) {
      const batch = chunks.slice(i, i + batchSize)
      const batchResults = await this.judgeBatch(query, batch)
      results.push(...batchResults)
    }

    return results
  }

  private async judgeBatch(query: string, chunks: IndexedChunk[]): Promise<JudgeResult[]> {
    const chunkDescriptions = chunks
      .map((c, idx) => {
        const preview = c.content.length > 300 ? c.content.slice(0, 297) + '...' : c.content
        return `[Chunk ${idx + 1}] ID: ${c.id}\nSource: ${c.source}\nSummary: ${c.summary}\nContent: ${preview}`
      })
      .join('\n\n')

    const prompt = `Query: "${query}"

Chunks to evaluate:

${chunkDescriptions}

Judge each chunk's relevance to the query. Return a JSON array.`

    try {
      const raw = await this.llm.generateJson<JudgeResult[]>(prompt, JUDGE_SYSTEM)

      if (!Array.isArray(raw)) {
        return this.fallbackJudge(chunks)
      }

      // Validate and normalize results
      return chunks.map((chunk) => {
        const found = raw.find((r) => r.id === chunk.id)
        if (found && typeof found.relevant === 'boolean' && typeof found.score === 'number') {
          return {
            id: chunk.id,
            relevant: found.relevant,
            score: Math.max(0, Math.min(1, found.score)),
          }
        }
        // If LLM didn't return this chunk's result, give it a neutral score
        return { id: chunk.id, relevant: false, score: 0.3 }
      })
    } catch {
      return this.fallbackJudge(chunks)
    }
  }

  /**
   * Fallback: give all chunks a neutral relevance score.
   * Better to return something than nothing.
   */
  private fallbackJudge(chunks: IndexedChunk[]): JudgeResult[] {
    return chunks.map((c) => ({
      id: c.id,
      relevant: true,
      score: 0.5,
    }))
  }
}
